File: hw2/codeforces/alarm.js
function alarmClock(input) {
    const params = input.split('\n');
    let [n, x, k] = params[0].split(' ').map(n => +n);
    let N = params[1].split(' ').map(n => +n);

    // let remainders = new Map();
    // for (let i = 0; i < n; i++) {
    //     let val = N[i];
    //     let key = val % x;
    //     let pval = remainders.get(key);

    //     if (pval !== undefined) {
    //         val = Math.min(pval, val);
    //     } 

    //     remainders.set(key, val);
    // }

    let R = new Set();
    N = N
        .sort((a, b) => a - b)
        .filter((a) => R.has(a % x) ? false : (R.add(a % x) || true))

    n = N.length;

    let bad = 0;
    let good = 1000000001000000001;

    while (good - bad > 1) {
        let mid = parseInt((bad + good) / 2);

        let count = 0;
        for (let i = 0; i < n; i++) {
            count += Math.max(parseInt(((mid - N[i]) / x) + 1), 0);
        }

        if (count >= k) good = mid;
        else bad = mid;
    }

    console.log(good);
}

let _input = "";
process.stdin.resume();
process.stdin.setEncoding("ascii");
process.stdin.on("data", (input) => _input += input);
process.stdin.on("end", () => alarmClock(_input));

File: hw2/codeforces/alarm.py
n, x, k = map(int, input().split())
N = list(map(int, input().split()))

d = dict()
for a in N:
    ai = a % x

    if (d.get(ai)):
        d[ai] = min(d[ai], a)
    else:
        d[ai] = a

A = d.values()

bad = 0
good = 1000000001000000001

while good - bad > 1:
    m = (bad + good) // 2
    c = 0

    for a in A:
        c += max(((m - a) // x) + 1, 0)

    if c >= k:
        good = m
    else:
        bad = m

print(good)
